lib/spack/spack/cmd/unload.py

```diff
@@ -4,20 +4,59 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
 import argparse
-from spack.cmd.common import print_module_placeholder_help
+import sys
 
-description = "remove package from environment using `module unload`"
-section = "modules"
+import llnl.util.tty as tty
+
+import spack.cmd
+import spack.environment as ev
+import spack.util.environment
+import spack.user_environment as uenv
+
+description = "remove package from the user environment variables`"
+section = "user environment"
 level = "short"
 
 
 def setup_parser(subparser):
     """Parser is only constructed so that this prints a nice help
        message with -h. """
+    shells = subparser.add_mutually_exclusive_group()
+    shells.add_argument(
+        '--sh', action='store_const', dest='shell', const='sh',
+        help="print sh commands to activate the environment")
+    shells.add_argument(
+        '--csh', action='store_const', dest='shell', const='csh',
+        help="print csh commands to activate the environment")
+
     subparser.add_argument(
-        'spec', nargs=argparse.REMAINDER,
+        'specs', nargs=argparse.REMAINDER,
         help='spec of package to unload with modules')
 
 
 def unload(parser, args):
-    print_module_placeholder_help()
+    env = ev.get_env(args, 'unload')
+    specs = list(map(lambda spec: spack.cmd.disambiguate_spec(spec, env),
+                     spack.cmd.parse_specs(args.specs)))
+    if not args.shell:
+        msg = [
+            "This command works best with Spack's shell support",
+            ""
+        ] + spack.cmd.common.shell_init_instructions + [
+            'Or, if you want to use `spack unload` without initializing',
+            'shell support, you can run one of these:',
+            '',
+            '    eval `spack unload --sh %s`   # for bash/sh' % args.specs,
+            '    eval `spack unload --csh %s`  # for csh/tcsh' % args.specs,
+        ]
+        tty.msg(*msg)
+        return 1
+
+    env_mod = spack.util.environment.EnvironmentModifications()
+    for spec in specs:
+        env_mod.extend(
+            uenv.environment_modifications_for_spec(spec).reversed())
+        env_mod.remove_path(uenv.spack_loaded_hashes_var, spec.dag_hash())
+    cmds = env_mod.shell_modifications(args.shell)
```

Comment:
In the current state, if you load something, then change the package so that the env mods would change, then unload the package, can any of these operations fail?  @alalazo?

I was worried about this and I think `EnvironmentModifications` is actually written so that things like `RemovePath` don't fail when the path to be removed is not present.  Specifically, `RemovePath.execute()` doesn't try to remove the thing that should be removed -- it just takes all the paths that are not the thing, which is great.  So, maybe this works fine and I do not need to worry.